Make network-simulator.py a little more useful

smfr opened the pull request:
#### b720c5d85f453afe369678698d8d44fb244f21f0
<pre>
Make network-simulator.py a little more useful
<a href="https://bugs.webkit.org/show_bug.cgi?id=268142">https://bugs.webkit.org/show_bug.cgi?id=268142</a>
<a href="https://rdar.apple.com/121642038">rdar://121642038</a>

Reviewed by Dewei Zhu and Alexey Proskuryakov.

Enhance http/tests/resources/network-simulator.py to support an initial response delay,
and variable chunk size, and chunk delay. This makes it possible to use to simulate
a slow-loading resource (potentially replacing the various &quot;slow.pl&quot; scripts we already have).

Add PDF as a valid resource type, and fix some typos.

A new test exercises the added functionality.

* LayoutTests/http/tests/harness/resources/chunked.txt: Added.
* LayoutTests/http/tests/harness/slow-loading-html-iframe.html: Added.
* LayoutTests/http/tests/resources/network-simulator.py:
(content_type):
(generate_response):
(file_to_stdout):
(chunked_file_to_stdout):
(handle_increase_resource_count_command):
(handle_increate_resource_count_command): Deleted.

Canonical link: <a href="https://commits.webkit.org/273607@main">https://commits.webkit.org/273607@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65b8efa9c00607e667e96403d571c2b1d90c31ca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35937 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14885 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38653 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32325 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17297 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11908 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31074 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36490 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12584 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31914 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11044 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11062 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32091 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39902 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32653 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37011 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11230 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9144 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35090 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12977 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31740 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8198 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11742 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12060 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov reviewed (approved):
Nice! Please get EWS to green (there was an infra failure with bot running out of space, but need to have the PR tested anyway).

dewei-zhu commented on LayoutTests/http/tests/resources/network-simulator.py:
```diff
@@ -74,20 +83,44 @@ def generate_response(path):
 
-def handle_increate_resource_count_command(path):
+def file_to_stdout(path):
+    if (chunk_delay):
```
Nit: parenthesis is not required.

dewei-zhu commented on LayoutTests/http/tests/resources/network-simulator.py:
```diff
@@ -74,20 +83,44 @@ def generate_response(path):
+        sys.stdout.buffer.write(open_file.read())
+
+
+def chunked_file_to_stdout(path, chunk_size, chunk_delay):
```
Nit: `chunk_size` and `chunk_delay` will shadow the global `chunk_size` and `chunk_delay`, maybe just call it `size` and `delay`?

dewei-zhu commented on LayoutTests/http/tests/resources/network-simulator.py:
```diff
@@ -74,20 +83,44 @@ def generate_response(path):
+            sys.stdout.buffer.write(chunk)
+            if chunk_delay:
+                time.sleep(chunk_delay / 1000)
+        sys.stdout.flush()
```
Why do we flush here but do not flush in `file_to_stdout` when `chunk_delay` is None?  This seems to be an inconsistency.

smfr replied:
Removed the final flush

dewei-zhu commented on LayoutTests/http/tests/resources/network-simulator.py:
```diff
@@ -146,6 +179,15 @@ def handle_log_resource_request(path):
+    initial_delay = int(initial_delay_arg)
+
+if chunk_size_arg is not None:
+    chunk_size = int(chunk_size_arg)
```
We may want to ensure `chunk_size != 0`, otherwise, file.read(0) is valid and will return an empty string per https://docs.python.org/3.8/library/io.html#io.RawIOBase.read

smfr replied:
I'll check for `chunk_size != 0` before using

dewei-zhu reviewed (approved):
r=me

smfr committed:
Make network-simulator.py a little more useful
https://bugs.webkit.org/show_bug.cgi?id=268142
rdar://121642038

Reviewed by Dewei Zhu and Alexey Proskuryakov.

Enhance http/tests/resources/network-simulator.py to support an initial response delay,
and variable chunk size, and chunk delay. This makes it possible to use to simulate
a slow-loading resource (potentially replacing the various "slow.pl" scripts we already have).

Add PDF as a valid resource type, and fix some typos.

A new test exercises the added functionality.

* LayoutTests/http/tests/harness/resources/chunked.txt: Added.
* LayoutTests/http/tests/harness/slow-loading-html-iframe.html: Added.
* LayoutTests/http/tests/resources/network-simulator.py:
(content_type):
(generate_response):
(file_to_stdout):
(chunked_file_to_stdout):
(handle_increase_resource_count_command):
(handle_increate_resource_count_command): Deleted.

Canonical link: https://commits.webkit.org/273607@main
